cloudprovider/aws/aws_provider.go

```diff
@@ -0,0 +1,90 @@
+package aws
+
+import (
+	"encoding/json"
+	"os"
+	"path"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/pkg/errors"
+	"github.com/rancher/agent/core/hostInfo"
+	"github.com/rancher/agent/utilities/config"
+)
+
+const (
+	regionLabel           = "ec2.region"
+	availabilityZoneLabel = "ec2.availability_zone"
+	infoFile              = "info.json"
+	tempFile              = "temp.json"
+)
+
+func GetCloudProviderInfo(expireTime time.Duration, interval time.Duration, getDocument func(*ec2metadata.EC2Metadata) (ec2metadata.EC2InstanceIdentityDocument, error)) bool {
```

Comment:
The way you setup this up to pass in the function is not bad...you are basically making use of dependency injection and mocks, which is good. And I like that the level of your mock is jsut getting the document from the client, so you don't have to mess with the actual client or an http server, but there are some improvements I'd like you to make:
1. By forcing the caller to pass in the function, you're exposing implementation details that should be hidden from the caller
2. You put the mock methods in this file, when they should really be in a _test.go file, so that they aren't compiled into the actual program. 

So, here's how we can improve:

in main.go:
```
awsProvider := aws.NewProvider()
go awsProvider.GetInfo()
```

in aws/aws_provider.go:
```
// Introduce a struct 
type provider struct {
  docClient documentClient
  expireTime
  interval
}

// documentClient is an interface
type documentClient infterface {
  getInstanceIdentityDocument
}

// the real impl of the interface is a struct with actual ec2 client on it
docClientImpl {
  client ec2metadataClient // or whatever this is called
}
(d *docClientImpl) func getInstanceIdentitydocument() {
  // call d.client.getInstanceIdentityDocument
}

// This is what main.go calls to get the struct:
func NewProvider() {
  // create a session, create a ec2 client, create a docClientImpl with the ec2 client, create a provider with the docClientImpl
  // return the docClientImpl
}
```

in your _test.go file:
```
// Add your mock implementations of documentClient here
// In your test, dont call NewProvider, create the provider struct directly and set the client as the mock client
```